test/distributed/fsdp/test_fsdp_comm_hooks.py

```diff
@@ -0,0 +1,172 @@
+# Owner(s): ["oncall: distributed"]
+
+import sys
+from typing import Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torch import distributed as dist
+from torch.distributed.algorithms.fsdp_comm_hooks import allreduce_hook
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp.fully_sharded_data_parallel import ShardingStrategy
+from torch.distributed.fsdp.wrap import (
+    enable_wrap,
+    wrap,
+)
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+class TestCommunicationHooks(FSDPTest):
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "has_wrapping", [True, False])
+    @parametrize(
+        "sharding_strategy",
+        [
+            ShardingStrategy.NO_SHARD,
+            ShardingStrategy.FULL_SHARD,
+            ShardingStrategy.SHARD_GRAD_OP
+        ])
+    def test_communication_hook_allreduce(
+        self,
+        has_wrapping: bool,
+        sharding_strategy: Optional[ShardingStrategy]
+    ):
+
+        """
+        Tests FSDP's all_reduce communication hook parity with default all_reduce.
+
+        Arguments:
+
+            has_wrapping (bool): Configures wrapping of a module.
+
+            sharding_strategy (Optional[ShardingStrategy]): Configures the FSDP algorithm.
+        """
+
+        class Net(nn.Module):
+
+            def __init__(self, has_wrapping):
+                # to ensure determinizm
+                torch.manual_seed(0)
+                torch.cuda.manual_seed(0)
+                super().__init__()
+                if has_wrapping:
+                    with enable_wrap(wrapper_cls=FSDP, device_id=torch.cuda.current_device()):
+                        self.fc1 = wrap(nn.Linear(8, 16))
+                else:
+                    self.fc1 = nn.Linear(8, 16)
+                self.relu = F.relu
```

Comment:
don't need this and just do F.relu in forward?

Reply:
Refactored